app/src/main/java/ai/elimu/appstore/BaseApplication.java

```diff
@@ -92,13 +102,55 @@ public Retrofit getRetrofit() {
         HttpLoggingInterceptor logging = new HttpLoggingInterceptor();
         logging.setLevel(BuildConfig.DEBUG ? HttpLoggingInterceptor.Level.BODY :
                 HttpLoggingInterceptor.Level.NONE);
-        final OkHttpClient okHttpClient = new OkHttpClient.Builder()
+
+        final ProgressResponseBody.ProgressListener progressListener = new ProgressResponseBody.ProgressListener() {
+            @Override
+            public void update(long bytesRead, long contentLength, boolean done) {
+//                Timber.i("bytesRead: " + bytesRead);
+//                Timber.i("contentLength: " + contentLength);
+//                Timber.i("done: " + done);
+//                Timber.i("%d%% done\n", (100 * bytesRead) / contentLength);
+
+                if (progressUpdateCallback != null) {
+                    long progress = (bytesRead * 100) / contentLength;
+//                    Timber.d("progress: " + progress);
+
+                    // E.g. "6.00 MB/12.00 MB   50%"
+                    String progressText = String.format(context.getString(R.string
+                                    .app_list_download_progress_number), bytesRead / 1024f / 1024f,
+                            contentLength / 1024f / 1024f, progress);
+
+//                    Timber.i("progressText: " + progressText);
+                    progressUpdateCallback.onProgressUpdated(progressText, (int) progress);
+                }
+            }
+        };
+
+
+        final OkHttpClient.Builder okHttpClientBuilder = new OkHttpClient.Builder()
                 .addInterceptor(logging)
                 .readTimeout(20, TimeUnit.SECONDS)
-                .retryOnConnectionFailure(true)
-                .build();
+                .retryOnConnectionFailure(true);
 
-        if (retrofit == null) {
+        /**
+         * In case of downloading/uploading big files and having necessity to update progress to UI, then add a
+         * progress listener
+         */
+        if (progressListener != null) {
+            okHttpClientBuilder.addNetworkInterceptor(new Interceptor() {
+                @Override
+                public Response intercept(Chain chain) throws IOException {
+                    Response originalResponse = chain.proceed(chain.request());
+                    return originalResponse.newBuilder()
+                            .body(new ProgressResponseBody(originalResponse.body(), progressListener))
+                            .build();
+                }
+            });
+        }
+
+        OkHttpClient okHttpClient = okHttpClientBuilder.build();
+
+        if (retrofit == null || progressUpdateCallback != null) {
```

Comment:
Why is Retrofit initialized if `progressUpdateCallback != null`?

Reply:
Each Retrofit instance is associated with a single OkHttpClient, which is also associated with only one single ProgressListener. 
When we have many concurrent downloads, each download should have a separate progress update callback. Therefore, it's necessary to initialize a new Retrofit connection for new download so that every download's progress is updated correctly.

When progressUpdateCallback is null, meaning that no progress update is required, then we can reuse existing Retrofit connection.

Reply:
Ok. Can you add this information to make the code easier to understand?

Reply:
Yep sure.